### PR TITLE
Various fixes to the differentiation module

### PR DESCRIPTION
**Pull request by BrunoLiegiBastonLiegi**
This implements some fixes to differentiation:
- Enables gradient wrt input for PSR
- Provides encoders with a ``differentiable`` property
- Improves ``Jax`` by using ``functools.partial`` instead of class attributes to avoid leakage

### CODE REVIEW

File: src/qiboml/models/encoding.py [hunk 1]
```diff
@@ -32,8 +35,14 @@ def circuit(
     ):
         return self._circuit
 
+    @property
+    def differentiable(self):
```
Comment (MatteoRobbiati) [inline src/qiboml/models/encoding.py:39]: Isn't maybe safer to set this property as False by default (1) or we assume cases like binary encoder are more exceptional while the default should in general be differentiable (2)?

File: src/qiboml/operations/differentiation.py [hunk 2]
```diff
@@ -49,11 +59,24 @@ def evaluate(self, x: ndarray, encoding, training, decoding, backend, *parameter
...
+                            encoding,
+                            circuit,
+                            decoding,
+                        )
+                    ),
+                    (decoding.output_shape[-1], x.shape[-1]),
+                )
+            )
+        else:
+            gradient.append(backend.np.zeros((decoding.output_shape[-1], x.shape[-1])))
```
Comment (MatteoRobbiati) [inline src/qiboml/operations/differentiation.py:79]: I am not sure we can just pad the gradient vector with values or zeros here. 
It is probably safe until the model structure is: encoding -> circuit -> decoding, but as soon as we will have a mix of trainable and non trainable gates we should probably loop over the gates. 
I say this because I guess the position of the gradients in the list is important, while here we are basically padding it at its beginning.

**Review by MatteoRobbiati — COMMENTED** [review]
Thanks again @BrunoLiegiBastonLiegi for this!! 
I just added some comments below. 
Let's discuss them quickly and - once this is done - for me we are good to go.

File: src/qiboml/interfaces/pytorch.py [hunk 3]
```diff
@@ -16,6 +16,7 @@
     "qiboml-pytorch": None,
     "qiboml-tensorflow": Jax,
     "qiboml-jax": Jax,
+    "numpy": Jax,
```
[the remark this answers is not in the document]
Comment (BrunoLiegiBastonLiegi) [reply src/qiboml/interfaces/pytorch.py:19]: yes since ``qibo`` is a dependency, and ``numpy`` its native backend.

File: src/qiboml/models/encoding.py (hunk 1 above)
Comment (BrunoLiegiBastonLiegi) [reply src/qiboml/models/encoding.py:39]: I would say it's more common to have a differentiable encoding.

File: src/qiboml/interfaces/pytorch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MatteoRobbiati) [reply src/qiboml/interfaces/pytorch.py:19]: > yes since `qibo` is a dependency, and `numpy` its native backend.

If Qibojit installed, the default should be Qibojit, no?

**Review by MatteoRobbiati — APPROVED** [review]
